### PR TITLE
Allow test clusters to run with TLS

### PR DESCRIPTION
**Pull request by scrawfor99**
### Description
Allow configuration of test clusters to run with TLS. 

I copied the change patterns from this PR: https://github.com/opensearch-project/OpenSearch/pull/1051/files. Thank you @VachaShah for the help. From there this is the linked issue about figuring out how to exercise the changes in these files directly: https://github.com/opensearch-project/OpenSearch/issues/1086. Seems unaddressed, so hopefully this will not be a blocker. 

### Related Issues

https://github.com/opensearch-project/security/issues/3056

### Check List
- [ ] ~New functionality includes testing.~
  - [ ] ~All tests pass~
- [x] New functionality has been documented.
  - [x] New functionality has javadoc added
- [x] Commits are signed per the DCO using --signoff
- [x] Commit changes are listed out in CHANGELOG.md file (See: [Changelog](../blob/main/CONTRIBUTING.md#changelog))

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: buildSrc/src/main/java/org/opensearch/gradle/testclusters/OpenSearchNode.java
Comment (peternied) [inline buildSrc/src/main/java/org/opensearch/gradle/testclusters/OpenSearchNode.java:1144]: Nit; Move this down by 1154

File: buildSrc/src/main/java/org/opensearch/gradle/testclusters/OpenSearchCluster.java
Comment (peternied) [inline buildSrc/src/main/java/org/opensearch/gradle/testclusters/OpenSearchCluster.java:1]: Broader question: There are tests cases like ReindexRestClientSslTests that seem to use a cluster with SSL client requests, is there a setup that is incorrect with the security plugin or is this a matter of the codepath at issue isn't used for those tests cases?

**Review by peternied — COMMENTED** [review]
I'm having trouble seeing how the value 'http.protocal' is used, shouldn't there be a change ~line 575 so the WaitForHttpResource object uses the correct protocol?

File: buildSrc/src/main/java/org/opensearch/gradle/testclusters/OpenSearchNode.java
Comment (scrawfor99) [reply buildSrc/src/main/java/org/opensearch/gradle/testclusters/OpenSearchNode.java:1151]: Sure

**Comment by scrawfor99** [comment]
> I'm having trouble seeing how the value 'http.protocal' is used, shouldn't there be a change ~line 575 so the WaitForHttpResource object uses the correct protocol?

Yeah, we need to take the value from the settings and use it where you mentioned. The issue is it is not clear how we can configure the tests on the fly and actually change that value. There is a lot of circular logic between the OpenSearchCluster and OpenSearchNode files and most the values appear hard coded. The settings that are passed are all about the file structure which is not what we want. I will have to ask someone more familiar with this test setup because I cannot find a way to actually pass a value on execution.

File: buildSrc/src/main/java/org/opensearch/gradle/testclusters/OpenSearchCluster.java
Comment (scrawfor99) [reply buildSrc/src/main/java/org/opensearch/gradle/testclusters/OpenSearchCluster.java:1]: I am not sure about this. I have really only looked at the two files: `OpenSearchNode` and `OpenSearchCluster`. I will take a look at the other tests to see if we can make use of those.

Comment (scrawfor99) [reply buildSrc/src/main/java/org/opensearch/gradle/testclusters/OpenSearchCluster.java:1]: It appears to setup a mock HTTP server for the requests:

```    @BeforeClass
    public static void setupHttpServer() throws Exception {
        InetSocketAddress address = new InetSocketAddress("localhost", 0);
        SSLContext sslContext = buildServerSslContext();
        server = HttpsServer.create(address, 0);
        server.setHttpsConfigurator(new ClientAuthHttpsConfigurator(sslContext));
        server.start();
        server.createContext("/", http -> {
            assert http instanceof HttpsExchange;
            HttpsExchange https = (HttpsExchange) http;
            handler.accept(https);
            // Always respond with 200
            // * If the reindex sees the 200, it means the SSL connection was established correctly.
            // * We can check client certs in the handler.
            https.sendResponseHeaders(200, 0);
            https.close();
        });
    }

```

File: buildSrc/src/main/java/org/opensearch/gradle/testclusters/OpenSearchNode.java
Comment (cwperks) [reply buildSrc/src/main/java/org/opensearch/gradle/testclusters/OpenSearchNode.java:1154]: `baseConfig` would be values that would be entered into `opensearch.yml` right?

With the security plugin, the setting that controls https is this one `plugins.security.ssl.http.enabled: true`. With testclusters is it possible to install a plugin and run a setup script? 

https://github.com/opensearch-project/security/blob/main/tools/install_demo_configuration.sh#L370

File: buildSrc/src/main/java/org/opensearch/gradle/testclusters/OpenSearchCluster.java
Comment (dblock) [inline buildSrc/src/main/java/org/opensearch/gradle/testclusters/OpenSearchCluster.java:376]: If I set the protocol to 'foobar' it will be ignored, I think you just want to read the `OpenSearchNode::getHttpProtocol` value and use that.

File: gradle/run.gradle
Comment (dblock) [inline gradle/run.gradle:43]: If `httpProto` is set, use that value.

**Review by dblock — CHANGES_REQUESTED** [review]
I think you should use the actual value set, if any. Today we support http/https, tomorrow we'll want to allow explicitly `http/2` or `grpc` or who knows what.

File: buildSrc/src/main/java/org/opensearch/gradle/testclusters/OpenSearchCluster.java
Comment (peternied) [reply buildSrc/src/main/java/org/opensearch/gradle/testclusters/OpenSearchCluster.java:1]: 👍 Thanks for looking into that works
👎 For it being useless

File: buildSrc/src/main/java/org/opensearch/gradle/testclusters/OpenSearchCluster.java
Comment (scrawfor99) [reply buildSrc/src/main/java/org/opensearch/gradle/testclusters/OpenSearchCluster.java:376]: That makes sense. I will swap over.

File: gradle/run.gradle
Comment (scrawfor99) [reply gradle/run.gradle:43]: Sounds good, I will swap over.

File: buildSrc/src/main/java/org/opensearch/gradle/testclusters/OpenSearchCluster.java
Comment (cwperks) [inline buildSrc/src/main/java/org/opensearch/gradle/testclusters/OpenSearchCluster.java:566]: Does this mean that a user testing bwc with security will first need to ensure that

```
plugins.security.ssl.http.enabled: true
```

to ensure that the node with the security plugin installed are running with `https` and then also set:

```
http.protocol: https
```

so that this single call will use `https` when making the request? Is there potentially another way of getting this setting from the plugin to control the scheme here or is that impossible?

Comment (scrawfor99) [reply buildSrc/src/main/java/org/opensearch/gradle/testclusters/OpenSearchCluster.java:566]: I believe so. We need this setup because we are trying to make the options configurable for two different purposes. On the Security Plugin side, we are only concerned with HTTP vs. HTTPS but with core, we need to handle HTTP vs. HTTPS or as DB mentioned other things like HTTP/2 etc.

**Comment by scrawfor99** [comment]
I don't know why the parsing is not working again. @dblock do you have any idea why the swap to the input would cause the failure even when it passed before? If not I will take another look, but I also do not know how to test this locally...

**Comment by scrawfor99** [comment]
Precommit failure is from runner crash. Should not be related to code.

**Comment by scrawfor99** [comment]
I believe this is all set.

**Review by peternied — DISMISSED** [review]

**Comment by scrawfor99** [comment]
Messaged GitHub on call for a second review.

**Review by cwperks — APPROVED** [review]

**Comment by scrawfor99** [comment]
I broke this again, because I think we need to be able to use the configuration setting to figure out the protocol for all nodes.

**Comment by peternied** [comment]
@scrawfor99 Were you able to test the changes locally?

**Comment by scrawfor99** [comment]
Hi @peternied, Craig figured out how to test the changes locally while I was out. I have not looked at what he did yet but I will be speaking with him about it this afternoon. We should be able to manually verify the changes we want to make as a result. 

That being said, I don't plan to add formal coverage tests as part of this PR given the testing scenario mentioned in the description.

**Comment by peternied** [comment]
I'm happy to approve once you've confirmed (manually or via automation) that this change will works.  I am concerned there might be an immediate follow up for how credentials are provided or trusting the certs from the cluster.

**Comment by scrawfor99** [comment]
Here are the steps to test locally:

1. Clone core and make all changes you want 
2. Clone security from this fork @cwperks made https://github.com/cwperks/security/tree/single-cluster-test
3. `./gradlew localDistro` & `./gradlew publishToMavenLocal` core 
4. Go to the location where the local distribution was placed and create a zip file of the distribution 
5. Go to the Security plugin and run `./gradlew assemble`
6. Grab the .zip file created in the security plugin and move it into the `single-cluster-test/src/test/resources/3.0.0.0` directory
7. Navigate back to `/single-cluster-test` and run ` ./gradlew runTestCluster -PcustomDistributionUrl="<path_to_core_zip>/opensearch-3.0.0-SNAPSHOT.tar.gz" -i`

This should allow you test changes by passing different settings with the `tests.opensearch.` prefix in build.gradle file of the single-cluster-test.

Unfortunately, this does not appear to work on M1 chips? 

I am presented a jarhell issue when running while @cwperks is able to run the exact same setup on an intel chip.

**Review by peternied — DISMISSED** [review]

**Comment by scrawfor99** [comment]
Just a follow-up, we don't want to merge this until testing is complete. The steps shared above work for Craig, but I am unable to use them on either my work or personal laptop.

**Comment by scrawfor99** [comment]
Update: The above approach seems to only work on Craig's laptop. Neither of us could reproduce it....

I found an alternative solution which involves using the System properties to pass a custom setting. 

By using command line we can get confirmation it works: ```Starting `node{::singleCluster0-1}`
SYSTEM PROPERTIES ARE: {java.specification.version=17, sun.jnu.encoding=UTF-8, sun.arch.data.model=64, java.vendor.url=https://adoptium.net/, sun.boot.library.path=/Users/steecraw/.sdkman/candidates/java/17.0.5-tem/lib, sun.java.command=org.gradle.launcher.daemon.bootstrap.GradleDaemon 8.2.1, jdk.debug=release, java.specification.vendor=Oracle Corporation, java.version.date=2022-10-18, java.home=/Users/steecraw/.sdkman/candidates/java/17.0.5-tem, file.separator=/, java.vm.compressedOopsMode=Zero based, line.separator=
, java.vm.specification.vendor=Oracle Corporation, java.specification.name=Java Platform API Specification, sun.management.compiler=HotSpot 64-Bit Tiered Compilers, java.runtime.version=17.0.5+8, user.name=steecraw, file.encoding=UTF-8, java.vendor.version=Temurin-17.0.5+8, java.io.tmpdir=/var/folders/8w/r3phsz2j09x9tky11pj8_ydh0000gr/T/, java.version=17.0.5, java.vm.specification.name=Java Virtual Machine Specification, tests.opensearch.http.protocol=https, native.encoding=UTF-8, java.library.path=/Users/steecraw/Library/Java/Extensions:/Library/Java/Extensions:/Network/Library/Java/Extensions:/System/Library/Java/Extensions:/usr/lib/java:., java.vendor=Eclipse Adoptium, sun.io.unicode.encoding=UnicodeBig, library.jansi.path=/Users/steecraw/.gradle/native/jansi/1.18/osx, java.class.path=/Users/steecraw/.gradle/wrapper/dists/gradle-8.2.1-bin/5hap6b9n41hkg4jeh2au2pllh/gradle-8.2.1/lib/gradle-launcher-8.2.1.jar, java.vm.vendor=Eclipse Adoptium, user.variant=, user.timezone=America/New_York, java.vm.specification.version=17, os.name=Mac OS X, user.country=US, sun.java.launcher=SUN_STANDARD, sun.cpu.endian=little, user.home=/Users/steecraw, user.language=en, org.gradle.appname=gradlew, path.separator=:, os.version=13.4, java.runtime.name=OpenJDK Runtime Environment, java.vm.name=OpenJDK 64-Bit Server VM, java.vendor.url.bug=https://github.com/adoptium/adoptium-support/issues, user.dir=/Users/steecraw/security/single-cluster-test, os.arch=aarch64, java.vm.info=mixed mode, java.vm.version=17.0.5+8, java.class.version=61.0}
Http protocol is now: https
``` 

We will also need to add a change to auth the request since without it we get an SSL exception.

**Comment by scrawfor99** [comment]
Security Plugin installs: 

Screenshot 2023-08-11 at 1 26 56 PM

Build Successful when running steps here: https://github.com/opensearch-project/OpenSearch/pull/8900#issuecomment-1672143707

**Comment by scrawfor99** [comment]
Command to run: 
```  ./gradlew runTestCluster -Dtests.opensearch.http.protocol=https -Dtests.opensearch.username=admin -Dtests.opensearch.password=admin -PcustomDistributionUrl="/Users/steecraw/OpenSearch/build/distribution/local/opensearch-3.0.0-SNAPSHOT.tar.gz" -i
```

File: buildSrc/src/main/java/org/opensearch/gradle/testclusters/OpenSearchCluster.java
Comment (peternied) [inline buildSrc/src/main/java/org/opensearch/gradle/testclusters/OpenSearchCluster.java:356]: Remove this log line?

File: buildSrc/src/main/java/org/opensearch/gradle/testclusters/OpenSearchNode.java
Comment (peternied) [inline buildSrc/src/main/java/org/opensearch/gradle/testclusters/OpenSearchNode.java:214]: Should we keep this log line?

**Review by peternied — APPROVED** [review]
Looks like spotless hit you, `./gradlew :build-tools:spotlessApply`.  Comments are nits

File: server/src/main/java/org/opensearch/env/Environment.java
Comment (cwperks) [inline server/src/main/java/org/opensearch/env/Environment.java:76]: Is this setting needed? The security plugin can pass test settings to testClusters' OpenSearchNode with `systemProperty <propertyKey> <propertyValue>` w/o adding any need cluster settings. This setting would also duplicate a setting from the security plugin which is used to determine whether `https` should be used.

File: buildSrc/src/main/java/org/opensearch/gradle/testclusters/OpenSearchCluster.java
Comment (scrawfor99) [reply buildSrc/src/main/java/org/opensearch/gradle/testclusters/OpenSearchCluster.java:356]: Will do

File: buildSrc/src/main/java/org/opensearch/gradle/testclusters/OpenSearchNode.java
Comment (scrawfor99) [reply buildSrc/src/main/java/org/opensearch/gradle/testclusters/OpenSearchNode.java:214]: We can remove it.

File: buildSrc/src/main/java/org/opensearch/gradle/testclusters/OpenSearchNode.java
Comment (scrawfor99) [reply buildSrc/src/main/java/org/opensearch/gradle/testclusters/OpenSearchNode.java:487]: Screenshot 2023-08-11 at 4 05 36 PM
You still have to use System.getProperty unfortunately. The class just affects how you pass things in not retrieve.

File: server/src/main/java/org/opensearch/env/Environment.java
Comment (scrawfor99) [reply server/src/main/java/org/opensearch/env/Environment.java:76]: I dont think so. Let me remove it.

Comment (scrawfor99) [reply server/src/main/java/org/opensearch/env/Environment.java:76]: Edit: We did not since we build each node with the system properties not from the configs. Removed.

**Comment by scrawfor99** [comment]
Flaky test failures: 

```


  | Test Result (6 failures  / +6)org.opensearch.cluster.routing.ShardMovementStrategyTests.testClusterGreenAfterPartialRelocationNoPreferenceShardMovementPrimaryFirstEnabledorg.opensearch.cluster.routing.ShardMovementStrategyTests.classMethodorg.opensearch.cluster.routing.allocation.BalancedSingleShardTests.classMethodorg.opensearch.cluster.routing.allocation.BalancedSingleShardTests.classMethodorg.opensearch.cluster.routing.allocation.MaxRetryAllocationDeciderTests.classMethodorg.opensearch.snapshots.SnapshotResiliencyTests.testTransportGetSnapshotsAction
-- | --


[Test Result](https://build.ci.opensearch.org/job/gradle-check/22531/testReport/) (6 failures / +6)

    [org.opensearch.cluster.routing.ShardMovementStrategyTests.testClusterGreenAfterPartialRelocationNoPreferenceShardMovementPrimaryFirstEnabled](https://build.ci.opensearch.org/job/gradle-check/22531/testReport/junit/org.opensearch.cluster.routing/ShardMovementStrategyTests/testClusterGreenAfterPartialRelocationNoPreferenceShardMovementPrimaryFirstEnabled/)
    [org.opensearch.cluster.routing.ShardMovementStrategyTests.classMethod](https://build.ci.opensearch.org/job/gradle-check/22531/testReport/junit/org.opensearch.cluster.routing/ShardMovementStrategyTests/classMethod/)
    [org.opensearch.cluster.routing.allocation.BalancedSingleShardTests.classMethod](https://build.ci.opensearch.org/job/gradle-check/22531/testReport/junit/org.opensearch.cluster.routing.allocation/BalancedSingleShardTests/classMethod/)
    [org.opensearch.cluster.routing.allocation.BalancedSingleShardTests.classMethod](https://build.ci.opensearch.org/job/gradle-check/22531/testReport/junit/org.opensearch.cluster.routing.allocation/BalancedSingleShardTests/classMethod_2/)
    [org.opensearch.cluster.routing.allocation.MaxRetryAllocationDeciderTests.classMethod](https://build.ci.opensearch.org/job/gradle-check/22531/testReport/junit/org.opensearch.cluster.routing.allocation/MaxRetryAllocationDeciderTests/classMethod/)
    [org.opensearch.snapshots.SnapshotResiliencyTests.testTransportGetSnapshotsAction](https://build.ci.opensearch.org/job/gradle-check/22531/testReport/junit/org.opensearch.snapshots/SnapshotResiliencyTests/testTransportGetSnapshotsAction/)
```

**Comment by peternied** [comment]
@scrawfor99 Retriggered CI

**Comment by peternied** [comment]
@scrawfor99 This test failure look related to your change
```
3: Task failed with an exception.
-----------
* What went wrong:
Execution failed for task ':example-plugins:custom-settings:yamlRestTest'.
> `cluster{:example-plugins:custom-settings:yamlRestTest}` failed to wait for cluster health yellow after 40 SECONDS
```

**Review by kkmr — APPROVED** [review]

**Comment by scrawfor99** [comment]
@peternied I think this should be set now

Flaky test:

```


  | Test Result (1 failure  / ±0)org.opensearch.client.RestClientSingleHostIntegTests.testRequestResetAndAbort
-- | --


[Test Result](https://build.ci.opensearch.org/job/gradle-check/22548/testReport/) (1 failure / ±0)

    [org.opensearch.client.RestClientSingleHostIntegTests.testRequestResetAndAbort](https://build.ci.opensearch.org/job/gradle-check/22548/testReport/junit/org.opensearch.client/RestClientSingleHostIntegTests/testRequestResetAndAbort/)
```

**Comment by scrawfor99** [comment]
Test failures are different: 

```


  | The following steps that have been detected may have insecure interpolation of sensitive variables         (click here for an explanation):         sh: [DOCKER_PASSWORD, DOCKER_USERNAME]
-- | --
 

<a href="https://build.ci.opensearch.org/job/gradle-check/22634/testReport/">Test Result</a> (2 failures  / ±0)<ul style="list-style-type: none; margin: 0;"><li style="" class="shown"><a href="https://build.ci.opensearch.org/job/gradle-check/22634/testReport/junit/org.opensearch.geo.search.aggregations.metrics/GeoBoundsITTestCase/classMethod/">org.opensearch.geo.search.aggregations.metrics.GeoBoundsITTestCase.classMethod</a></li><li style="" class="shown"><a href="https://build.ci.opensearch.org/job/gradle-check/22634/testReport/junit/org.opensearch.index.translog/RemoteFSTranslogTests/testMetadataFileDeletion/">org.opensearch.index.translog.RemoteFSTranslogTests.testMetadataFileDeletion</a></li></ul>The following steps that have been detected may have insecure interpolation of sensitive variables [(click here for an explanation)](https://jenkins.io/redirect/groovy-string-interpolation):

    sh: [DOCKER_PASSWORD, DOCKER_USERNAME]

[Test Result](https://build.ci.opensearch.org/job/gradle-check/22634/testReport/) (2 failures / ±0)

    [org.opensearch.geo.search.aggregations.metrics.GeoBoundsITTestCase.classMethod](https://build.ci.opensearch.org/job/gradle-check/22634/testReport/junit/org.opensearch.geo.search.aggregations.metrics/GeoBoundsITTestCase/classMethod/)
    [org.opensearch.index.translog.RemoteFSTranslogTests.testMetadataFileDeletion](https://build.ci.opensearch.org/job/gradle-check/22634/testReport/junit/org.opensearch.index.translog/RemoteFSTranslogTests/testMetadataFileDeletion/)
```


Seems like definitely flaky.

**Comment by scrawfor99** [comment]
@dblock can I get a review and possible merge please? Thank you :)

File: buildSrc/src/main/java/org/opensearch/gradle/testclusters/OpenSearchNode.java
Comment (reta) [inline buildSrc/src/main/java/org/opensearch/gradle/testclusters/OpenSearchNode.java:213]: Why do we need to add an empty map here?

File: buildSrc/src/main/java/org/opensearch/gradle/testclusters/TestClusterConfiguration.java
Comment (scrawfor99) [reply buildSrc/src/main/java/org/opensearch/gradle/testclusters/TestClusterConfiguration.java:110]: Previously dB had suggested swapping it from just being based on HTTP to allowing a more generic string entry: https://github.com/opensearch-project/OpenSearch/pull/8900#pullrequestreview-1550809454

File: buildSrc/src/main/java/org/opensearch/gradle/testclusters/OpenSearchNode.java
Comment (scrawfor99) [reply buildSrc/src/main/java/org/opensearch/gradle/testclusters/OpenSearchNode.java:213]: Adding an empty map properly instantiates the object `this.credentials`. Because `this.credentials` is defined as `private final List<Map<String, String>> credentials = new ArrayList<>();` then we had a list of maps with no maps in it. This meant the straightforward logic here 

```
   if (System.getProperty("tests.opensearch.username") != null) {
            this.credentials.get(0).put("username", System.getProperty("tests.opensearch.username"));
            LOGGER.info("Overwriting username to: " + this.getCredentials().get(0).get("username"));
        }
        if (System.getProperty("tests.opensearch.password") != null) {
            this.credentials.get(0).put("password", System.getProperty("tests.opensearch.password"));
            LOGGER.info("Overwriting password to: " + this.getCredentials().get(0).get("password"));
        }
```
would require duplicate instantiation (if only one was set we could get NPE)  or a separate instantiation loop. I am not sure why this is a list of maps instead of just a map but this made the most sense to me.

File: buildSrc/src/main/java/org/opensearch/gradle/testclusters/TestClusterConfiguration.java
Comment (reta) [reply buildSrc/src/main/java/org/opensearch/gradle/testclusters/TestClusterConfiguration.java:110]: Sure, but there are 2 part to that:
 - protocol (let say, http/1.1, http/2, http/3, grpc)
 - TLS / DTLS

```
http/1.1 + TLS = https
http/2 + TLS = https
http/3 + TLS = https
```

So we need two properties here I think: protocol + security.

Comment (scrawfor99) [reply buildSrc/src/main/java/org/opensearch/gradle/testclusters/TestClusterConfiguration.java:110]: Sorry, I am not sure I follow. Does providing the end result (https or http) not allow for the same configuration? Sorry if I am misunderstanding.

Comment (scrawfor99) [reply buildSrc/src/main/java/org/opensearch/gradle/testclusters/TestClusterConfiguration.java:111]: Does the swap I made address your concern? Letting us input a string allows for more configuration in how we want to test things it seems. For instance we can confirm DTLS and TLS work the same but may bea ble to later add tests that examine how they differ. This seems like what you were suggesting originally.

Comment (reta) [reply buildSrc/src/main/java/org/opensearch/gradle/testclusters/TestClusterConfiguration.java:111]: > Does the swap I made address your concern?

It goes towards right direction for sure but since this is a string, what value(s) should be acceptable there? (boolean flag is significantly easier to reason about)

Comment (scrawfor99) [reply buildSrc/src/main/java/org/opensearch/gradle/testclusters/TestClusterConfiguration.java:111]: I will swap it to a boolean but you won't be able to swap between UDP and TCP.

File: buildSrc/src/main/java/org/opensearch/gradle/testclusters/OpenSearchCluster.java
Comment (reta) [inline buildSrc/src/main/java/org/opensearch/gradle/testclusters/OpenSearchCluster.java:565]:
```suggestion
            String resolvedProtocol = getFirstNode().getHttpProtocol(), getFirstNode().isSecure() ? "https" : "http";
```

File: buildSrc/src/main/java/org/opensearch/gradle/testclusters/OpenSearchCluster.java
Comment (scrawfor99) [reply buildSrc/src/main/java/org/opensearch/gradle/testclusters/OpenSearchCluster.java:632]: Why do we have the first string then?

File: buildSrc/src/main/java/org/opensearch/gradle/testclusters/OpenSearchNode.java
Comment (scrawfor99) [reply buildSrc/src/main/java/org/opensearch/gradle/testclusters/OpenSearchNode.java:501]: Sorry but I am not sure what is being asked... 

I am not sure why we added two different settings--it seems like we are going back to an effective boolean between HTTP and HTTPS which was what I originally was asked not to use.

**Review by reta — APPROVED** [review]

**Comment by reta** [comment]
Thanks @scrawfor99 ! We needs signoff from @dblock

**Comment by scrawfor99** [comment]
Failing precommit seems unrelated.

**Review by dblock — APPROVED** [review]
Code looks good. So, how do I use this?

**Comment by scrawfor99** [comment]
Hi @dblock, that is a good question haha. I am adding some notes to using it in the Security repo, but basically it is used with this process: https://github.com/opensearch-project/security/issues/3056#issuecomment-1675386278. In short, with this change, you can go into the security repo and grab the code changes that will shortly be merged when I fix them all. The working branch is here: https://github.com/scrawfor99/security/tree/testClusterChanges.

Then you will run the BWC tests `bwcTestSuite` and provide arguments for the system properties I detailed here. Specifically, it will look something like 

```
./gradlew -p bwc-test clean bwcTestSuite \
   -Dtests.security.manager=false \
   -Dtests.opensearch.http.protocol=https \
   -Dtests.opensearch.username=admin \
   -Dtests.opensearch.password=admin \
   -PcustomDistributionUrl="/Users/steecraw/OpenSearch/distribution/archives/linux-tar/build/distributions/opensearch-min-3.0.0-SNAPSHOT-linux-x64.tar.gz" \
   -i
```

This lets you specify security information when running our BWC tests so we can then confirm when the tests operate with Security and establish secure connections between the nodes.

**Comment by dblock** [comment]
Can we add this info to https://github.com/opensearch-project/OpenSearch/blob/main/TESTING.md#testing-backwards-compatibility as part of this PR please?

File: buildSrc/src/main/java/org/opensearch/gradle/testclusters/OpenSearchCluster.java
Comment (Yury-Fridlyand) [inline buildSrc/src/main/java/org/opensearch/gradle/testclusters/OpenSearchCluster.java:578]: Why do you set credentials twice here?

File: buildSrc/src/main/java/org/opensearch/gradle/testclusters/OpenSearchCluster.java
Comment (Yury-Fridlyand) [inline buildSrc/src/main/java/org/opensearch/gradle/testclusters/OpenSearchCluster.java:567]: Why `useradd` and `-p` and not `username` and `password`?

**Review by Yury-Fridlyand — COMMENTED** [review]
Sorry for the late review. Your changes broke some of my tests in another repo and I started investigation.

File: buildSrc/src/main/java/org/opensearch/gradle/testclusters/OpenSearchCluster.java
Comment (scrawfor99) [reply buildSrc/src/main/java/org/opensearch/gradle/testclusters/OpenSearchCluster.java:578]: One constructs the URL, one configures the the object.

File: buildSrc/src/main/java/org/opensearch/gradle/testclusters/OpenSearchCluster.java
Comment (scrawfor99) [reply buildSrc/src/main/java/org/opensearch/gradle/testclusters/OpenSearchCluster.java:567]: This was support for the previous testing

**Comment by scrawfor99** [comment]
> Sorry for the late review. Your changes broke some of my tests in another repo and I started investigation.

What repo? Do you have an issue you can link?

**Comment by Yury-Fridlyand** [comment]
@scrawfor99 
I already found a fix. But you are welcome to review:
https://github.com/Bit-Quill/opensearch-project-sql/pull/335

**Comment by acarbonetto** [comment]
@scrawfor99 updated link for Yury's PR: https://github.com/opensearch-project/sql/pull/1986
Please take a look if you have a moment.  This allows us to run IT tests against a cluster & remote cluster that has the security and SQL plugins installed.

**Pull request merged**
